### DISCUSSION TITLE
Logging/Alert System

### DISCUSSION THREAD

**Discussion by rwinstanley1** [alan-turing-institute/data-safe-haven#1249] [Ideas]
#### :scroll: Description
Linked to issue alan-turing-institute/data-safe-haven-production#40 as part of requirements to have a system

- 6.1.1
  - system to identify data security and protection breaches
  - to work with the DSP Toolkit incident reporting tool?
- 6.1.4
  - potentially an alert to those responsible for data security built into the system?

requirements don't specify that for either of these it has to be a technical solution, but, was discussed previously that there was a desire for a technical logging solution by the team so think it's worth looking into. could be a combination of technical and manual processes for March (especially for the 'alert' part of requirement). 

#### :strawberry: Desired behaviour
To have an (ideally) technical solution for logging and alerts. 

#### :page_facing_up: Tasks
Include specific tasks (if any) in the order in which they need to be done.
- [ ] Identify potential logging options that could be deployed in existing infrastructure. 
- [ ] Find out if there is a way to automate system alerts at least partly, if not what would manual process look like?

**Comment by JimMadge** [comment]
#### Identify potential logging options that could be deployed in existing infrastructure.

Currently we use [Azure Log Analytics](https://docs.microsoft.com/en-us/azure/azure-monitor/log-query/log-analytics-overview) to query logs in [Azure Monitor](https://docs.microsoft.com/en-us/azure/azure-monitor/overview)(?). Feedback is that querying data is not very straightforward. How hard is it to integrate non-Azure things into this system, for example generic Linux logs in `/var/log` or logs in a container?

Alternatives worthconsidering,

- [Graylog](https://www.graylog.org/products/open-source)
- [Logstash](https://www.elastic.co/logstash)
- [journalcheck](https://github.com/lynix/journalcheck)
- [logwatch](https://sourceforge.net/projects/logwatch/)

#### Find out if there is a way to automate system alerts at least partly, if not what would manual process look like?

A fully automated system seems unlikely, how can we automatically determine what should raise concern?

With one of the dashboard type solutions we could configure searches/dashboards to help filter useful information (_e.g._ authentication logs, antivirus logs). These should also let us create alerts for events which obviously merit investigation (_e.g._ antivirus positive results).

**Comment by JimMadge** [comment]
There is a question here of what needs to be logged. For example, we could (presumably) extract text and journald logs from the containers running internal webapps (GitLab, HackMD). I expect this would take more effort than getting similar data from the host OS on a VM and it feels like this information is less important. Due to the isolation model of the safe haven, if there is malicious activity on one of these webapps then the safe haven has already be compromised. (But maybe we should be more worried about people breaking out?)

Similarly, I expect we don't want to keep logs of users command history.

**Comment by JimMadge** [comment]
## Graylog

- Ingest logs from [journald, syslog, Windows eventlog, text, JSON via HTTP and others](https://docs.graylog.org/en/3.3/pages/sending_data.html#log-sources)
- Uses elastic search
- [Ansible role!](https://galaxy.ansible.com/Graylog2/graylog-ansible-role) and docker image options
- [Alerts](https://docs.graylog.org/en/3.3/pages/alerts.html#alerts-events)
- It looks like we might need to use [Azure Monitor](https://azure.microsoft.com/en-us/blog/use-azure-monitor-to-integrate-with-siem-tools/) to interface with Azure logs (via Log Analytics)

**Comment by jemrobinson** [comment]
I think Docker can log to `/var/log/syslog` on the host OS. Also, I'm happy to take another look at `auditd` for ensuring that user-initiated events are logged in the syslog on the DSVM.

**Comment by JimMadge** [comment]
## Plan

### Technical

- Implement a centralised logging system using Graylog
- Log _important_ information to the centralised system
  - DSVM activity
  - Malware alerts (clamav)
  - Internet facing services
- Alternatively, log more information than is necessary and rely on filters to find what is important/relevant
- Create alerts for obvious problems using Graylog API

### Policy template

- The `<safe haven admin>` should respond to any alerts immediately by contacting the `<person with overall responsibility for data security>`
- Additionally the `<safe haven admin>` should regularly (every `<1> <days/weeks>`) inspect the items on the logging system for any problems and raise any issues as above

**Comment by JimMadge** [comment]
## Current scope

- Deploy graylog server to SHM and configure for ingesting syslog
- Create 'events' on graylog for sudo access and clamav threats
- Configure DSVMs to send syslog data to graylog server

## Future work

- Create email alerts (requires an SMTP email account)
- Ingest windows logs using https://docs.graylog.org/en/4.0/pages/sending/windows.html
- Ingest Azure logs using https://github.com/miguelangelopereira/azuremonitor2syslog as Azure function (using beats input for graylog)

**Comment by martintoreilly** [comment]
@JimMadge @jemrobinson Can someone spend a decent amount of time working with the Azure logging and log analytics / monitoring / reporting services to see if these can meet our needs? I feel we've not spent much serious effort on this yet. This feels like a good area to collaborate with closely with IT colleagues on, as they may well already be using some of these services to monitor our enterprise systems on Azure.

**Comment by martintoreilly** [comment]
In the GrayLog context, can we write to Azure storage / database with append only rights and access these for analysis / reporting / monitoring / alerting with read only rights?

**Comment by martintoreilly** [comment]
Looks like we can [configure custom logs via Powershell](https://docs.microsoft.com/en-us/azure/azure-monitor/platform/powershell-workspace-configuration) for the Azure log analytics workspace.

**Comment by jemrobinson** [comment]
@martintoreilly There are certainly some logs (especially those to do with the Azure fabric/infrastructure) that are difficult to capture except with LogAnalytics. However, the main advantage of GrayLog is that it's not dependent on Azure which makes it easier to port to AWS (which is important for several potential endusers of this setup).

It looks like we should be able to [forward Azure Monitor Logs](https://github.com/miguelangelopereira/azuremonitor2syslog) to GrayLog using an Azure function. If this works then I'm reasonably comfortable in using GrayLog for log aggregation/alerting and possibly providing a useful dashboard which can ingest data:

- directly from Linux and Windows VMs
- indirectly from LogAnalytics for Azure-only components

This seems like the easiest way to plan for a multi-platform future.

**Comment by martintoreilly** [comment]
@jemrobinson Ok. I totally get the cross-platform argument, especially if it's also less / no more work that getting the Azure-only option working. My main concern is that ability to delete / edit logs is as restricted as possible, ideally to no more than the group who can deploy / delete infrastructure. Can we configure the underlying datastore for GrayLog so that only infrastructure admins can edit / delete log data and within-safe-haven-admins, local root on the Gray Log box or local root on any box contributing logging data to Gray Log can't?

**Comment by JimMadge** [comment]
# Azure Monitor & Azure Sentinel

## Azure Monitor
- Log aggregation (Monitor), querying (Log Analytics), dashboards and alerts
- Azure resource
- Requires log analytics agent on hosts ([supports](https://docs.microsoft.com/en-us/azure/azure-monitor/platform/agents-overview#supported-operating-systems) Ubuntu 18.04, 20.04 Windows Server 2016, 2019) **or** log analytics [VM extension](https://docs.microsoft.com/en-us/azure/virtual-machines/extensions/oms-linux?toc=/azure/azure-monitor/toc.json)
- Agent forwards
  - syslogs on Linux hosts 
  - Windows logs
  - VM Performance
- ~([Installation instructions](https://docs.microsoft.com/en-us/azure/azure-monitor/platform/agent-linux#install-the-agent-using-wrapper-script), log analytics details/credentials are given at install time)~ Agent/VM extension can be installed via Powershell. See [Setup_SRE_Logging.ps1](https://github.com/alan-turing-institute/data-safe-haven/blob/master/deployment/secure_research_environment/setup/Setup_SRE_Logging.ps1) and [Deployments.psm1#L845](https://github.com/alan-turing-institute/data-safe-haven/blob/aa24fc74ee2a2d6873230ccec172d2f5db24a242/deployment/common/Deployments.psm1#L845) for how we currently install the extension on SRE VMs.
- Custom logs are definable via Powershell (see [example is docs](https://docs.microsoft.com/en-us/azure/azure-monitor/platform/powershell-workspace-configuration))
- Logs are sent via HTTPS (over the internet? It is unclear whether this is only for 'on-prem') https://docs.microsoft.com/en-us/azure/azure-monitor/platform/log-analytics-agent#firewall-requirements
- Log analytics workspace must have globally unique name 🤯 


## Sentinel
- security information event management (SIEM) (focused on searching, visualising and querying logs rather than collecting)
- $2.46 per GB ingested pay-as-you-go pricing

**Comment by martintoreilly** [comment]
[Setup_SRE_Logging.ps1](https://github.com/alan-turing-institute/data-safe-haven/blob/master/deployment/secure_research_environment/setup/Setup_SRE_Logging.ps1) also used to configure the collection of various Windows and Linux performance counters and activation of Log Analytics insight features, but these seem to be missing from master. I can find them in [commit 76127bd](https://github.com/alan-turing-institute/data-safe-haven/blob/76127bd455f944e85def4c4fad2b99bb2a7e4f52/deployment/safe_haven_management_environment/setup/Setup_SHM_Logging.ps1) ~and they seem to be present when PR alan-turing-institute/data-safe-haven#877 was merged~.

@jemrobinson @JimMadge Either of you know where we lost this?

Edit: It looks like it didn't actually make it into the original PR alan-turing-institute/data-safe-haven#743 merge where the changes were made, though I'm struggling to figure out why. The lines are definitely present in commit 198d960 via the "commits" tab of PR alan-turing-institute/data-safe-haven#743, but don't appear at all in the history of [Setup_SRE_Logging.ps1](https://github.com/alan-turing-institute/data-safe-haven/commits/98e1790dd4707d2c7f3acd82e99dcae228d6b139/deployment/secure_research_environment/setup/Setup_SRE_Logging.ps1) (going back from commit 98e1790) via the "files" tab of PR alan-turing-institute/data-safe-haven#743.

**Comment by jemrobinson** [comment]
@martintoreilly I think this is in [Setup_SHM_Logging.ps1](https://github.com/alan-turing-institute/data-safe-haven/blob/master/deployment/safe_haven_management_environment/setup/Setup_SHM_Logging.ps1)

**Comment by martintoreilly** [comment]
> @martintoreilly I think this is in [Setup_SHM_Logging.ps1](https://github.com/alan-turing-institute/data-safe-haven/blob/master/deployment/safe_haven_management_environment/setup/Setup_SHM_Logging.ps1)

Yep. There it is. Thanks. I thought I was going mad.

My eyes weren't working and I didn't spot that it was the SHM logging setup file rather than the SRE one in commit 198d960.

**Comment by jemrobinson** [comment]
@JimMadge Here are some Microsoft docs that might be useful: [how to bring syslog data into LogAnalytics](https://docs.microsoft.com/en-us/azure/azure-monitor/platform/data-sources-syslog) and [overview of Azure Monitor agents](https://docs.microsoft.com/en-us/azure/azure-monitor/platform/agents-overview).

**Comment by JimMadge** [comment]
# Ingesting Azure Monitor logs into Graylog

The scheme set out by the [Forward Azure Monitor Logs to Syslog](https://github.com/miguelangelopereira/azuremonitor2syslog) action looks fairly complicated. It involves,

- Azure Monitor
- Azure EventHub
- Function service
- Syslog server (in our case Graylog)

**Comment by JimMadge** [comment]
# Steps required to ingest syslog from Linux VMs

- Deploy log analytics workspace and VM
- Create entries for each syslog [facility](https://en.wikipedia.org/wiki/Syslog#Facility) (note that names are significant), enabling all severities, in log analytics workspace (seems only automatable through [powershell](https://docs.microsoft.com/en-us/azure/azure-monitor/platform/powershell-workspace-configuration))
- [Install Azure Monitor for VMs](https://docs.microsoft.com/en-us/azure/virtual-machines/extensions/oms-linux?toc=/azure/azure-monitor/toc.json) using log analytics workspace name and primary (or secondary) key to link VMs to the workspace

Rough powershell,

```powershell
$ResourceGroup = ""
$WorkspaceName = ""
$WorkspaceId = ""
$WorkspaceKey = ""
$VMName = ""
$Location = ""
$FacilityNames       = @()
$FacilityNames      += 'auth'
$FacilityNames      += 'authpriv'
$FacilityNames      += 'cron'
$FacilityNames      += 'daemon'
$FacilityNames      += 'ftp'
$FacilityNames      += 'kern'
$FacilityNames      += 'mail'
$FacilityNames      += 'syslog'
$FacilityNames      += 'user'
$FacilityNames      += 'uucp'

$Count = 0
foreach ($FacilityName in $FacilityNames) {
    $Count++
    $null = New-AzOperationalInsightsLinuxSyslogDataSource `
    -ResourceGroupName $ResourceGroupName `
    -WorkspaceName $WorkspaceName `
    -Name "Linux-syslog-$($Count)" `
    -Facility $FacilityName `
    -CollectEmergency `
    -CollectAlert `
    -CollectCritical `
    -CollectError `
    -CollectWarning `
    -CollectNotice `
    -CollectDebug `
    -CollectInformational
}

$Settings = @{
    "workspaceId" = $WorkspaceId
}
$ProtectedSettings = @{
    "workspaceKey" = $WorkspaceKey
}

Set-AzVMExtension -Name OMSLinux -ExtensionType OmsAgentForLinux -Publisher Microsoft.EnterpriseCloud.Monitoring -TypeHandlerVersion "1.13" -ResourceGroupName $ResourceGroup -VMName $VMName -Location $Location -Settings $Settings -ProtectedSettings $ProtectedSettings
```

**Comment by martintoreilly** [comment]
## Azure monitor private endpoint
There's an [Azure Monitor Private Endpoint](https://docs.microsoft.com/en-gb/azure/azure-monitor/logs/private-link-security) that ensures two-way restriction of DSH / Monitor instance (i.e. our DSH can only connect to our Monitor instance and vice-versa) and that traffic to the various Azure monitor service endpoints runs over the Azure internal network.

Credit: [Microsoft's Azure TRE design](https://microsoft.github.io/AzureTRE/azure-tre-overview/networking/) uses this.
